src/main/java/io/confluent/connect/jdbc/dialect/GenericDatabaseDialect.java

```diff
@@ -0,0 +1,1753 @@
+      Connection conn = connections.poll();
+      if (conn != null) {
+        try {
+          conn.close();
+        } catch (SQLException e) {
+          log.warn("Error while closing connection to {}", jdbcDriverInfo, e);
+        }
+      }
+    }
+  }
+
+  @Override
+  public boolean isConnectionValid(
+      Connection connection,
+      int timeout
+  ) throws SQLException {
+    if (jdbcDriverInfo().jdbcMajorVersion() >= 4) {
+      return connection.isValid(timeout);
+    }
+    // issue a test query ...
+    String query = checkConnectionQuery();
+    if (query != null) {
+      try (Statement statement = connection.createStatement()) {
+        boolean hasResultSet = statement.execute(query);
+        if (hasResultSet) {
+          statement.getResultSet().close();
+        }
+      }
+    }
+    return true;
+  }
+
+  /**
+   * Return a query that can be used to check the validity of an existing database connection
+   * when the JDBC driver does not support JDBC 4. By default this returns {@code SELECT 1},
+   * but subclasses should override this when a different query should be used.
+   *
+   * @return the check connection query; may be null if the connection should not be queried
+   */
+  protected String checkConnectionQuery() {
+    return "SELECT 1";
+  }
+
+  protected JdbcDriverInfo jdbcDriverInfo() {
+    if (jdbcDriverInfo == null) {
+      try (Connection connection = getConnection()) {
+        jdbcDriverInfo = createJdbcDriverInfo(connection);
+      } catch (SQLException e) {
+        throw new ConnectException("Unable to get JDBC driver information", e);
+      }
+    }
+    return jdbcDriverInfo;
+  }
+
+  protected JdbcDriverInfo createJdbcDriverInfo(Connection connection) throws SQLException {
+    DatabaseMetaData metadata = connection.getMetaData();
+    return new JdbcDriverInfo(metadata.getJDBCMajorVersion(), metadata.getJDBCMinorVersion(),
+                              metadata.getDriverName(), metadata.getDatabaseProductName(),
+                              metadata.getDatabaseProductVersion()
+    );
+  }
+
+  /**
+   * Add or modify any connection properties based upon the {@link #config configuration}.
+   *
+   * <p>By default this method does nothing and returns the {@link Properties} object supplied as a
+   * parameter, but subclasses can override it to add/remove properties used to create new
+   *
+   * @param properties the properties that will be passed to the {@link DriverManager}'s {@link
+   *                   DriverManager#getConnection(String, Properties) getConnection(...) method};
+   *                   never null
+   * @return the updated connection properties, or {@code properties} if they are not modified or
+   *     should be returned; never null
+   */
+  protected Properties addConnectionProperties(Properties properties) {
+    return properties;
+  }
+
+  @Override
+  public PreparedStatement createPreparedStatement(
+      Connection db,
+      String query
```

Comment:
should this method be here? It seems like it should be in a JDBC utils package/class than in a dialect implementation.

Reply:
It does seem strange, but yes. This gives dialects the ability to set fetch direction, query timeouts, max rows for the result sets, etc.

Reply:
Also, IMO it does not make sense to use a static JDBC utility method for things like this, since doing so prevents the ability customize the behavior -- that's what this PR is trying to move away from in the first place.

Now, one might argue that simply creating a prepared statement is something so basic that you shouldn't need anything but the JDBC API. And when you're working with a single DBMS that is definitely true. However, we're dealing with _lots_ of driver implementations and DBMSes, and they all have their quirks, limitations, and pain points. IMO, we _have_ to be prepared to customize almost all interaction with the JDBC driver.